pgx/src/memcxt.rs

```diff
@@ -174,28 +174,40 @@ pub enum PgMemoryContexts {
 
 /// A `pg_sys::MemoryContext` that is owned by `PgMemoryContexts::Owned`
 #[derive(Debug)]
-pub struct OwnedMemoryContext(pg_sys::MemoryContext);
+pub struct OwnedMemoryContext {
+    owned: pg_sys::MemoryContext,
+    previous: pg_sys::MemoryContext,
+}
 
 impl Drop for OwnedMemoryContext {
     fn drop(&mut self) {
         unsafe {
-            pg_sys::MemoryContextDelete(self.0);
+            // In order to prevent crashes, if we're trying to drop
+            // a context that is current, switch to its predecessor, and then drop it
+            if pg_sys::CurrentMemoryContext == self.owned {
```

Comment:
I think this probably should use [`std::ptr::eq()`](https://doc.rust-lang.org/std/ptr/fn.eq.html).  Just a little more explicit that we are comparing pointers.